runtime/lua/vim/fs.lua

```diff
@@ -86,6 +86,66 @@ function M.basename(file)
   return file:match('/$') and '' or (file:match('[^/]*$'))
 end
 
+--- @class vim.fs.abspath.Opts
+--- @inlinedoc
+---
+--- Current working directory to use as the base for the relative path.
+--- This option is ignored for C:foo\bar style paths on Windows, as those paths use the current
+--- directory of the drive specified in the path.
+--- (default: |current-directory|)
+--- @field cwd? string
+
+--- Convert path to an absolute path. Does not check if the path exists, resolve symlinks or
+--- hardlinks (including `.` and `..`), or expand environment variables. If the path is already
+--- absolute, it is returned unchanged.
+---
+--- @param path string Path
+--- @param opts vim.fs.abspath.Opts? Optional keyword arguments:
+--- @return string Absolute path
+function M.abspath(path, opts)
```

Comment:
if `paths` is a list, that has a few advantages:

- allows multiple segments to be combined, similar to nodejs resolve() https://nodejs.org/api/path.html#pathresolvepaths
- when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:
    - `abspath('foo', {cwd: '.'})`
    - `abspath({'.', 'foo'})`
    - making it explicit has other advantages, such as reminder plugin authors to think about `getcwd()` and window-local paths.

Reply:
> * allows multiple segments to be combined, similar to nodejs resolve()

`vim.fs.joinpath()` can do that already

> * when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:when calling, the "cwd" is the first arg, which is more readable and more explicit. compare:

Tbh with you, I personally do not find it more intuitive, the opposite, even. And even if it were more intuitive, in the vast majority of cases, you're going to not set `cwd` manually, and in those cases, just having a single path argument is much simpler.

Reply:
Alternatively could define it as `path: string|table`. That allows opt-in to explicit multi-arg form. Although that's a bit redundant with joinpath(), it eliminates the `opts.cwd` option.

> in the vast majority of cases, you're going to not set cwd manually, and in those cases, just having a single path argument is much simpler.

For the single-arg case we can default to using CWD as the root dir.

Reply:
> Alternatively could define it as `path: string|table`. That allows opt-in to explicit multi-arg form. Although that's a bit redundant with joinpath(), it eliminates the `opts.cwd` option.
> 
> > in the vast majority of cases, you're going to not set cwd manually, and in those cases, just having a single path argument is much simpler.
> 
> For the single-arg case we can default to using CWD as the root dir.

We might need to add more opts (like path OS style and whether to expand env variables) to `abspath` in the future, so imo keeping the opts dict is best for now, it's also more in line with the abspath function in something like Python.  Having a dict just adds more complexity for the majority of usecases

Reply:
> We might need to add more opts (like path OS style and whether to expand env variables) to `abspath` in the future,

Irrelevant. The door is open for that in the future. But today, we can support a list arg as the first argument, that's natural and more readable than a CWD arg that _follows_ the first arg.

Reply:
I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`, especially since the use case here is already recognised to be quite rare.

Reply:
> I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`

Because it avoids the `cwd` option.

Reply:
> > I'm not sure why we need a list argument when the user can simply use `vim.fs.joinpath`
> 
> Because it avoids the `cwd` option. If we can avoid both, then that's fine too.

I would personally pick not having `cwd` over using a list argument. The list argument imo is not really intuitive. Here's how I see it:

`abspath` is a function that takes a path and returns the absolute path. What is the absolute path for a table supposed to mean? Imo it can really trip up new users. If removing `cwd` is the solution, then I'm fine with doing that

Reply:
For the record, I'm with famiu here. I don't understand the objection to the original signature at all, _provided_ that there's a compelling use case for resolving paths relative to non-current directories. (Maybe `cwd` is the wrong name and `base` or similar would be better?) If we currently don't have such a use case -- either in our own code or a reasonably common plugin -- we can leave it for later (with the understanding that we'll simply add it and _not_ start fighting over it again).

Reply:
Should we rather implement this as an option on `normalize()`? Not sure that `abspath()` is worth all the extra docs.

Reply:
Apples and oranges. `normalize()` does not need a cwd, since its action doesn't depend on it. Only the conversion relative to absolute path needs to know what the path is relative _to_ .

The thing to keep in mind is that you may wish to normalize a relative path, or make a non-normalized path absolute. So it's either small, composable, single-purpose functions (the *ooonix way*) or one big swiss army knife function that transforms paths, controlled by (a bunch of) options you need to track interactions of.

Reply:
> The thing to keep in mind is that you may wish to normalize a relative path, or make a non-normalized path absolute. 

What I meant was, similar to the `expand_env` option, normalize() could have a `abs` option. 

> So it's either small, composable, single-purpose functions (the _ooonix way_) or one big swiss army knife function that transforms paths, controlled by (a bunch of) options you need to track interactions of.

Yes, but the 2 main concepts are "normalize" and "resolve". Currently `normalize()` is also doing some "resolving" (the `expand_env` option), and arguably that should be pulled into a separate `resolve()` or `expand()` function.

Reply:
Anyway to be clear, this PR is not blocked on the above.